### DISCUSSION TITLE
ado.query 参数查询问题

### DISCUSSION THREAD

**Discussion by s899000** [dotnetcore/FreeSql#1299] [Q&A]
var query = DB.RoadFlow.Ado
                .Query<RoadFlow.Data.Model.Dictionary>("SELECT * FROM Dictionary WHERE ParentID=(SELECT ID FROM Dictionary WHERE Code=@Code) ORDER BY Sort",
                new { Code = code });

![image](https://user-images.githubusercontent.com/33239560/195540068-70a8eddc-8b6b-4808-b19d-f8a55953b2ed.png)
![image](https://user-images.githubusercontent.com/33239560/195540269-a239362f-22ca-4a93-84aa-45062f97ec98.png)
数据库有数据，orm查询没有数据

**Comment by 2881099** [comment]
连错数据库

**Reply by s899000** [reply]
不是啊，哥，我怎么可能犯这种低级错误，mysql的

**Reply by 2881099** [reply]
试试不用 @code 参数化，直接写 sql 值。查询 count(1)，看到回结果

**Comment by s899000** [comment]
![image](https://user-images.githubusercontent.com/33239560/195605621-db965de1-ef26-443e-938a-92291ba3ed42.png)
![image](https://user-images.githubusercontent.com/33239560/195605681-683144e8-c387-42aa-b6f2-e509bf65e018.png)
自己看吧

**Reply by s899000** [reply]
nuget里安装的freesql

**Reply by 2881099** [reply]
https://freesql.net/guide/ado.html
看下参数化说明

**Reply by 2881099** [reply]
Ado 下面所有参数 object parms 都可以接受匿名对象，或者字典：

- new { id = 1, name = "xx" }
- new Dictionary\<string, object\> { ["id"] = 1, ["name"] = "xx" }

关于参数前缀：

- odbc 是 ? 并且没有标识，所以 freesql 禁用了 odbc 参数化
- oracle 是 :
- mysql.data 是 ?
- mysqlconnector 是 @
- 其他基本都是 @

IN 参数化查询：

> 当前仅支持 Array 和 IList 类型绑定

```csharp
var ids = new int[] { 1,2,3 };
List<T> list = fsql.Ado.Query<T>("select * from t1 where id in @ids", new { ids = ids });
```

**Reply by 2881099** [reply]
FreeSql.Provider.MySql 和 FreeSql.Provider.MySqlConnector 的区别

https://freesql.net/guide/freesql-provider-mysqlconnector.html

**Reply by 2881099** [reply]
Ado 是辅助功能，不是 FreeSql 核心功能，编写 SQL 需要了解对应的 adonet 驱动语法。
